src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,324 @@
+  // Map keys are absolute paths, while map values are paths that are
+  // relative to the output path.
+  map<string, string> output_path_aliases = 6;
+}
+
+message StartBuildResponse {
+  // If set, the contents of the output path are almost entirely identical on
+  // the results of a previous build. This information may be used by Bazel to
+  // prevent unnecessary scanning of the file system.
+  //
+  // The server MUST leave this field unset in case the contents of the output
+  // path are empty, not based on a previous build, if no tracking of this
+  // information is performed, or if the number of changes made to the output
+  // path is too large to be expressed.
+  InitialOutputPathContents initial_output_path_contents = 1;
+
+  // A path that the client must append to StartBuildRequest.output_path_prefix
+  // to obtain the full path at which outputs of the build are stored.
+  //
+  // Bazel replaces bazel-out/ with a symlink targeting this path.
+  string output_path_suffix = 2;
+}
+
+message InitialOutputPathContents {
+  // The identifier of a previously finalized build whose results are stored in
+  // the output path.
+  string build_id = 1;
+
+  // Output files that have been modified (either the content or metadata) or
+  // removed since the build was finalized.
+  //
+  // If directories as a whole are modified/removed, the server needs to
+  // recursively list all modified paths underneath.
+  //
+  // If the remote output service freezes the contents of the output files after
+  // FinalizeArtifacts, this field can be left empty.
+  repeated string modified_files = 2;
+}
+
+message StageArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Describe how to stage the artifact.
+    google.protobuf.Any locator = 2;
+  }
+
+  // The identifier of the build. Server uses this to determine which output
+  // path needs to be modified.
+  string build_id = 1;
+
+  repeated Artifact artifacts = 2;
+}
+
+message StageArtifactsResponse {
+  message Response {
+    // If the status has a code other than `OK`, it indicates that the artifact
+    // cannot be staged.
+    //
+    // Errors:
+    // * `NOT_FOUND`: The requested Artifact is not in the CAS.
+    google.rpc.Status status = 1;
+  }
+
+  // The response for each of the requested artifacts, using the same order as
+  // requested. This means that this list has the same length as
+  // StageArtifactsRequest.artifacts.
+  repeated Response responses = 1;
+}
+
+message FinalizeArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Expected digest for this file. This allows server to detect if the file
+    // has been changed after Bazel finished creating the file and the
+    // corresponding FinalizeArtifactsRequest is processed. This field is
+    // optional for regular files, and should be omitted for directories and
```

Comment:
Should the SHOULD be a MUST? (I think making it optional for directories and symlinks invites questions that we'd prefer not to answer.)

Reply:
Changed to MUST.